add  optimal parameter configrations. (#5303)
pingcap/docs-cn: check-before-deployment.md

sudo systemctl enable ntpd.service
```

## 检查和配置操作系统优化参数

在生产系统的 TiDB 中，建议对操作系统进行如下的优化配置：

1. 关闭透明大页（即 Transparent Huge Pages，缩写为 THP）。数据库的内存访问模式往往是稀疏的而非连续的，而且当高阶内存碎片化比较严重时，分配 THP 页面会出现较大的延迟。

> 2. 将存储介质的 IO 调度器设置为 noop。对于高速 SSD 存储介质，内核的 IO 调度操作会导致性能损失。调度器设置为 noop，内核不做任何操作，尽可能快地将 IO 请求下发给硬件，能获取更好的性能。同时，noop 调度器也有较好的普适性。

Reviewer: + `1. 2. 3` 这三步是否存在时间先后顺序呢？
+ `将调度器设置为 noop` 、`不对内核做任何操作`、`尽可能快地将 I/O 请求下发给硬件` 三者是并列的操作吗？
```suggestion
2. 将存储介质的 I/O 调度器设置为 noop。对于高速 SSD 存储介质，内核的 I/O 调度操作会导致性能损失。将调度器设置为 noop ，不对内核做任何操作，尽可能快地将 I/O 请求下发给硬件，能获取更好的性能。同时，noop 调度器也有较好的普适性。
```

> 2. 将存储介质的 I/O 调度器设置为 noop。对于高速 SSD 存储介质，内核的 I/O 调度操作会导致性能损失。将调度器设置为 noop ，不对内核做任何操作，尽可能快地将 I/O 请求下发给硬件，能获取更好的性能。同时，noop 调度器也有较好的普适性。

Reviewer: ```suggestion
2. 将存储介质的 I/O 调度器设置为 noop。对于高速 SSD 存储介质，内核的 I/O 调度操作会导致性能损失。将调度器设置为 noop 后，调度器不对内核做任何操作，可尽快将 I/O 请求下发给硬件，以获取更好的性能。同时，noop 调度器也有较好的普适性。
```

Author: 后，内核不做任何操作，尽可能快地将

> 3. 调整 CPU 频率的 cpufreq 模块选用 performance 模式。将 CPU 频率固定工作在其支持的最高运行频率上，不进行动态调节，以获取最佳的性能。

Reviewer: ```suggestion
3. 调整 CPU 频率的 cpufreq 模块选用 performance 模式。将 CPU 频率固定在其支持的最高运行频率上，不进行动态调节，以获取最佳的性能。
```

> 3. 调整 CPU 频率的 cpufreq 模块选用 performance 模式。将 CPU 频率固定在其支持的最高运行频率上，不进行动态调节，以获取最佳的性能。

Reviewer: ```suggestion
3. 为调整 CPU 频率的 cpufreq 模块选用 performance 模式。将 CPU 频率固定在其支持的最高运行频率上，不进行动态调节，可获取最佳的性能。
```

> 采用如下步骤检查操作系统的当前配置，并配置操作系统优化参数：

Reviewer: ```suggestion
采用如下步骤检查操作系统的当前配置，并配置系统优化参数：
```

    [always] madvise never
    ```

    > **注意：**
    >
    > `[always] madvise never` 表示透明大页处于启用状态，需要进行关闭。

2. 执行以下命令查看数据目录所在磁盘的 IO 调度器。

>    假设有 sdb、sdc 两个磁盘上创建了数据目录。

Reviewer 2: ```suggestion
   假设在 sdb、sdc 两个磁盘上创建了数据目录。
```

    {{< copyable "shell-regular" >}}

    ```bash
    cat /sys/block/sd[bc]/queue/scheduler
    ```

    ```
    noop [deadline] cfq
    noop [deadline] cfq
    ```
    
    > **注意：**
    >
    > `noop [deadline] cfq` 表示磁盘的 IO 调度器使用 deadline，需要进行修改。

3. 执行以下命令查看磁盘的唯一标识 ID_SERIAL。

    {{< copyable "shell-regular" >}}

    ```bash
    udevadm info --name=/dev/sdb | grep ID_SERIAL
    ```

    ```
    E: ID_SERIAL=36d0946606d79f90025f3e09a0c1f9e81
    E: ID_SERIAL_SHORT=6d0946606d79f90025f3e09a0c1f9e81
    ```
    
    > **注意：**
    >
    > 如果有多个磁盘都分配了数据目录，需要执行多次，记录所有磁盘的唯一标识。

4. 执行以下命令查看 cpufreq 模块选用的节能策略。

    {{< copyable "shell-regular" >}}

    ```bash
    cpupower frequency-info --policy
      ```

    ```
    analyzing CPU 0:
    current policy: frequency should be within 1.20 GHz and 3.10 GHz.
                  The governor "powersave" may decide which speed to use within this range.
    ```
    
    > **注意：**
    >
    > `The governor "powersave"` 表示cpufreq 的节能策略使用 powersave，需要进行修改。虚拟机或者云主机不需要调整，命令输出通常为 `Unable to determine current policy` 。

> 5. 推荐使用 tuned 方式配置应用优化参数。

Reviewer 2: ```suggestion
5. 配置应用优化参数方式一：tuned 方式（推荐）。
```

        - network-latency             - Optimize for deterministic performance at the cost of increased power consumption, focused on low latency network performance
        - network-throughput          - Optimize for streaming network throughput, generally only necessary on older CPUs or 40G+ networks
        - powersave                   - Optimize for low power consumption
        - throughput-performance      - Broadly applicable tuning that provides excellent performance across a variety of common server workloads
        - virtual-guest               - Optimize for running inside a virtual guest
        - virtual-host                - Optimize for running KVM guests
        Current active profile: balanced
        ```

        `Current active profile: balanced` 表示当前操作系统的 tuned 策略使用 balanced，建议在当前策略的基础上添加操作系统优化配置。

    2. 创建新的 tuned 策略。

        {{< copyable "shell-regular" >}}

        ```bash
        mkdir /etc/tuned/balanced-tidb-optimal/
        vi /etc/tuned/balanced-tidb-optimal/tuned.conf
        ```

        ```
        [main]
        include=balanced

        [cpu]
        governor=performance

        [vm]
        transparent_hugepages=never

        [disk]

>         devices_udev_regex=(ID_SERIAL=36d0946606d79f90025f3e09a0c1fc035)|(ID_SERIAL=36d0946606d79f90025f3e09a0c1f9e81)

Reviewer 3: 有个疑问，如果用户因磁盘损坏等问题换盘，tuned 的行为是怎样的？

Author: 用 ID_SERIAL 的话，需要重新编辑，因为正则匹配不上了。之前 ORACLE 用 udev 规则写磁盘阵列上的 LUN 的权限时，有时候会用 ID_VENDOR 把厂商 ID 给绑定。在我们的服务器内部磁盘的场景中，可能没有办法用上这种方法。

        elevator=noop
        ```

        `include=balanced` 表示在现有的 balanced 策略基础上添加操作系统优化配置。
    
    3. 应用新的 tuned 策略。

        {{< copyable "shell-regular" >}}

        ```bash
        tuned-adm profile balanced-tidb-optimal
        ```
    

> 6. 使用脚本的配置方式应用优化参数。如果已经使用 tuned 方式，请跳过本步骤。

Reviewer 2: ```suggestion
6. 配置应用优化参数方式二：使用 `grubby` 脚本方式，如果已经使用 tuned 方式，请跳过本步骤。
```

    [always] madvise never
    ```

    > **注意：**
    >

>     > `[always] madvise never` 表示透明大页处于启用状态，需要进行关闭。

Reviewer: ```suggestion
    > `[always] madvise never` 表示透明大页处于启用状态。需要关闭透明大页。
```

Author: ，需要关闭。

2. 执行以下命令查看数据目录所在磁盘的 I/O 调度器。假设在 sdb、sdc 两个磁盘上创建了数据目录。

    {{< copyable "shell-regular" >}}

    ```bash
    cat /sys/block/sd[bc]/queue/scheduler
    ```

    ```
    noop [deadline] cfq
    noop [deadline] cfq
    ```
    
    > **注意：**
    >

>     > `noop [deadline] cfq` 表示磁盘的 IO 调度器使用 `deadline`，需要进行修改。

Reviewer: 需要将 [deadline] 修改为什么呢？
```suggestion
    > `noop [deadline] cfq` 表示磁盘的 I/O 调度器使用 `deadline`，需要进行修改。
```

3. 执行以下命令查看磁盘的唯一标识 `ID_SERIAL`。

    {{< copyable "shell-regular" >}}

    ```bash
    udevadm info --name=/dev/sdb | grep ID_SERIAL
    ```

    ```
    E: ID_SERIAL=36d0946606d79f90025f3e09a0c1f9e81
    E: ID_SERIAL_SHORT=6d0946606d79f90025f3e09a0c1f9e81
    ```
    
    > **注意：**
    >

>     > 如果有多个磁盘都分配了数据目录，需要多次执行以上命令，记录所有磁盘的唯一标识。

Reviewer: ```suggestion
    > 如果多个磁盘都分配了数据目录，需要多次执行以上命令，记录所有磁盘各自的唯一标识。
```

4. 执行以下命令查看 cpufreq 模块选用的节能策略。

    {{< copyable "shell-regular" >}}

    ```bash
    cpupower frequency-info --policy
    ```

    ```
    analyzing CPU 0:
    current policy: frequency should be within 1.20 GHz and 3.10 GHz.
                  The governor "powersave" may decide which speed to use within this range.
    ```
    
    > **注意：**
    >

>     > `The governor "powersave"` 表示 cpufreq 的节能策略使用 powersave，需要进行调整。如果是虚拟机或者云主机，则不需要调整，命令输出通常为 `Unable to determine current policy`。

Reviewer: ```suggestion
    > `The governor "powersave"` 表示 cpufreq 的节能策略使用 powersave，需要调整为其他策略。如果是虚拟机或者云主机，则不需要调整，命令输出通常为 `Unable to determine current policy`。
```

Author: 需要调整为 performance 策略.

Reviewer: 已改

    [always] madvise never
    ```

    > **注意：**
    >
    > `[always] madvise never` 表示透明大页处于启用状态，需要进行关闭。

> 2. 执行以下命令查看数据目录所在磁盘的 IO 调度器。
>    假设在 sdb、sdc 两个磁盘上创建了数据目录。

Reviewer: ```suggestion
2. 执行以下命令查看数据目录所在磁盘的 I/O 调度器。假设在 sdb、sdc 两个磁盘上创建了数据目录。
```

    {{< copyable "shell-regular" >}}

    ```bash
    cat /sys/block/sd[bc]/queue/scheduler
    ```

    ```
    noop [deadline] cfq
    noop [deadline] cfq
    ```
    
    > **注意：**
    >

>     > `noop [deadline] cfq` 表示磁盘的 IO 调度器使用 deadline，需要进行修改。

Reviewer: ```suggestion
    > `noop [deadline] cfq` 表示磁盘的 IO 调度器使用 `deadline`，需要进行修改。
```

> 3. 执行以下命令查看磁盘的唯一标识 ID_SERIAL。

Reviewer: ```suggestion
3. 执行以下命令查看磁盘的唯一标识 `ID_SERIAL`。
```

    {{< copyable "shell-regular" >}}

    ```bash
    udevadm info --name=/dev/sdb | grep ID_SERIAL
    ```

    ```
    E: ID_SERIAL=36d0946606d79f90025f3e09a0c1f9e81
    E: ID_SERIAL_SHORT=6d0946606d79f90025f3e09a0c1f9e81
    ```
    
    > **注意：**
    >

>     > 如果有多个磁盘都分配了数据目录，需要执行多次，记录所有磁盘的唯一标识。

Reviewer: ```suggestion
    > 如果有多个磁盘都分配了数据目录，需要多次执行以上命令，记录所有磁盘的唯一标识。
```

4. 执行以下命令查看 cpufreq 模块选用的节能策略。

    {{< copyable "shell-regular" >}}

    ```bash
    cpupower frequency-info --policy

>       ```

Reviewer: ```suggestion
    ```
```

    ```
    analyzing CPU 0:
    current policy: frequency should be within 1.20 GHz and 3.10 GHz.
                  The governor "powersave" may decide which speed to use within this range.
    ```
    
    > **注意：**
    >

>     > `The governor "powersave"` 表示cpufreq 的节能策略使用 powersave，需要进行修改。虚拟机或者云主机不需要调整，命令输出通常为 `Unable to determine current policy` 。

Reviewer: ```suggestion
    > `The governor "powersave"` 表示 cpufreq 的节能策略使用 powersave，需要进行调整。如果是虚拟机或者云主机，则不需要调整，命令输出通常为 `Unable to determine current policy`。
```
